README.md
# Topobuf

[![Build Status](https://app.travis-ci.com/azavea/topobuf.svg?branch=main)](https://travis-ci.org/azavea/topobuf)

Topobuf is a compact binary encoding for topographic data.

Topobuf provides _lossless_[^1] compression of [TopoJSON](https://github.com/topojson/topojson) data into [protocol buffers](https://developers.google.com/protocol-buffers/).
Advantages over using JSON-based formats alone:

- **Very compact**: typically makes TopoJSON 2-3 times smaller.
- Smaller even when comparing gzipped sizes: 20-30% for TopoJSON.
- Can store topology objects too large for `JSON.stringify` / `JSON.parse`
- Can accommodate any TopoJSON data, including extensions with arbitrary properties.

#### Sample compression sizes
|                       | normal   | gzipped |
|-----------------------|----------|---------|
| pa-census-blocks.json | 280 MB   | 44 MB   |
| pa-census-blocks.pbf  | 114 MB   | 35 MB   |
| us-zips.json          | 15.02 MB | 3.19 MB |
| us-zips.pbf           | 4.85 MB  | 2.72 MB |
| idaho.json            | 1.9 MB   | 612 KB  |
| idaho.pbf             | 567 KB   | 479 KB  |


## Install

```bash
npm install topobuf
```

## API

### encode

```js
var buffer = topobuf.encode(topojson, new Pbf());
```

Given a TopoJSON object and a [Pbf](https://github.com/mapbox/pbf) object to write to,
returns a Topobuf as a `Buffer` object in Node or `UInt8Array` object in browsers.

### decode

```js
var topojson = topobuf.decode(new Pbf(data));
```

Given a [Pbf](https://github.com/mapbox/pbf) object with topobuf data, return a TopoJSON object.

## See more

This library is based on [`geobuf`](https://github.com/mapbox/geobuf) by Mapbox, which provides similar functionality for GeoJSON

[^1]: When using [quantized TopoJSON](https://github.com/topojson/topojson-client/blob/master/README.md#quantize) - _nearly lossless_ otherwise

CHANGELOG.md
## 1.1.0 (Feb 23, 2022)

Performance improvements to decode function

## 1.0.0 (Feb 18, 2022)

Library forked from [`geobuf`](https://github.com/mapbox/geobuf) version `v1.0.1`

test/validate.test.js
var topobuf = require("../index"),
  Pbf = require("pbf"),
  test = require("tap").test,
  fs = require("fs");

var files = ["airports", "multi-line", "props", "no-transform", "simple", "us-states"];
for (var i = 0; i < files.length; i++) {
  test(
    "roundtrip TopoJSON: " + files[i],
    roundtripTest(getJSON(files[i] + ".topo.json"))
  );
}

function roundtripTest(geojson) {
  return function (t) {
    var resp = topobuf.decode(new Pbf(topobuf.encode(geojson, new Pbf())));
    t.same(resp, geojson);
    t.end();
  };
}

function getJSON(name) {
  return JSON.parse(fs.readFileSync(__dirname + "/fixtures/" + name));
}


decode.js
'use strict';

module.exports = decode;

var keys, values, lengths, dim, e, transformed, names, prevP, valuesIndex;

var geometryTypes = ['Point', 'MultiPoint', 'LineString', 'MultiLineString',
                      'Polygon', 'MultiPolygon', 'GeometryCollection'];

function decode(pbf) {
    dim = 2;
    e = Math.pow(10, 6);
    transformed = false;
    lengths = null;

    keys = [];
    values = [];
    valuesIndex = 0;
    var obj = pbf.readFields(readDataField, {});
    keys = null;
    values = null;

    return obj;
}

function readDataField(tag, obj, pbf) {
    if (tag === 1) keys.push(pbf.readString());
    else if (tag === 2) dim = pbf.readVarint();
    else if (tag === 3) e = Math.pow(10, pbf.readVarint());

    else if (tag === 4) readTopology(pbf, obj);
}

function readGeometry(pbf, geom) {
    return pbf.readMessage(readGeometryField, geom);
}

function readTopology(pbf, topology) {
    topology.type = 'Topology';
    topology.objects = {};
    topology.arcs = [];
    names = [];
    prevP = new Array(dim);
    pbf.readMessage(readTopologyField, topology);
    names = null;
    return topology;
}

function readTopologyField(tag, topology, pbf) {
    if (tag === 1) {
        topology.transform = pbf.readMessage(readTransformField, {scale: [], translate: []});
        transformed = true;
    }
    else if (tag === 2) names.push(pbf.readString());
    else if (tag === 3) topology.objects[names.shift()] = pbf.readMessage(readGeometryField, {});

    else if (tag === 4) lengths = pbf.readPackedVarint();
    else if (tag === 5) topology.arcs = readArcs(pbf);

    else if (tag === 13) values[valuesIndex++] = (readValue(pbf));
    else if (tag === 15) readProps(pbf, topology);
}

function readGeometryField(tag, geom, pbf) {
    if (tag === 1) geom.type = geometryTypes[pbf.readVarint()];

    else if (tag === 2) lengths = pbf.readPackedVarint();
    else if (tag === 3) readCoords(geom, pbf, geom.type);
    else if (tag === 4) {
        geom.geometries = geom.geometries || [];
        geom.geometries.push(readGeometry(pbf, {}));
    }

    else if (tag === 11) geom.id = pbf.readString();
    else if (tag === 12) geom.id = pbf.readSVarint();

    else if (tag === 13) values[valuesIndex++] = (readValue(pbf));
    else if (tag === 14) geom.properties = readProps(pbf, {});
    else if (tag === 15) readProps(pbf, geom);
}

function readCoords(geom, pbf, type) {
    if (type === 'Point') geom.coordinates = readPoint(pbf);
    else if (type === 'MultiPoint') geom.coordinates = readLine(pbf, true);
    else if (type === 'LineString') geom.arcs = readLine(pbf);
    else if (type === 'MultiLineString' || type === 'Polygon') geom.arcs = readMultiLine(pbf);
    else if (type === 'MultiPolygon') geom.arcs = readMultiPolygon(pbf);
}

function readValue(pbf) {
    var end = pbf.readVarint() + pbf.pos,
        value = null;

    while (pbf.pos < end) {
        var val = pbf.readVarint(),
            tag = val >> 3;

        if (tag === 1) value = pbf.readString();
        else if (tag === 2) value = pbf.readDouble();
        else if (tag === 3) value = pbf.readVarint();
        else if (tag === 4) value = -pbf.readVarint();
        else if (tag === 5) value = pbf.readBoolean();
        else if (tag === 6) value = JSON.parse(pbf.readString());
    }
    return value;
}

function readProps(pbf, props) {
    var end = pbf.readVarint() + pbf.pos;
    while (pbf.pos < end) props[keys[pbf.readVarint()]] = values[pbf.readVarint()];
    valuesIndex = 0;
    return props;
}

function readTransformField(tag, tr, pbf) {
    if (tag === 1) tr.scale[0] = pbf.readDouble();
    else if (tag === 2) tr.scale[1] = pbf.readDouble();
    else if (tag === 3) tr.translate[0] = pbf.readDouble();
    else if (tag === 4) tr.translate[1] = pbf.readDouble();
}

function readPoint(pbf) {
    var end = pbf.readVarint() + pbf.pos,
        coords = [];
    while (pbf.pos < end) coords.push(transformCoord(pbf.readSVarint()));
    return coords;
}

function readLinePart(pbf, end, len, isMultiPoint) {
    var i = 0,
        coords = len ? new Array(len) : [],
        p, d;

    if (!isMultiPoint) {
        p = 0;
        while (len ? i < len : pbf.pos < end) {
            p += pbf.readSVarint();
            coords[i] = p;
            i++;
        }

    } else {
        for (d = 0; d < dim; d++) prevP[d] = 0;

        while (len ? i < len : pbf.pos < end) {
            p = new Array(dim);
            for (d = 0; d < dim; d++) {
                prevP[d] += pbf.readSVarint();
                p[d] = transformCoord(prevP[d]);
            }
            coords[i] = p;
            i++;
        }
    }

    return coords;
}

function readLine(pbf, isMultiPoint) {
    return readLinePart(pbf, pbf.readVarint() + pbf.pos, null, isMultiPoint);
}

function readMultiLine(pbf) {
    var end = pbf.readVarint() + pbf.pos;
    if (!lengths) return [readLinePart(pbf, end)];

    var len = lengths.length,
        coords = new Array(len);
    for (var i = 0; i < len; i++) coords[i] = readLinePart(pbf, end, lengths[i]);
    lengths = null;
    return coords;
}

function readMultiPolygon(pbf) {
    var end = pbf.readVarint() + pbf.pos;
    if (!lengths) return [[readLinePart(pbf, end)]];

    var len = lengths[0];
    var coords = new Array(len);
    var j = 1;
    for (var i = 0; i < len; i++) {
        var rlen = lengths[j],
            rings = new Array(rlen);
        for (var k = 0; k < rlen; k++) rings[k] = readLinePart(pbf, end, lengths[j + 1 + k]);
        j += lengths[j] + 1;
        coords[i] = rings;
    }
    lengths = null;
    return coords;
}

function readArcs(pbf) {
    var len = lengths.length,
        lines = new Array(len),
        end = pbf.readVarint() + pbf.pos;

    for (var i = 0; i < len; i++) {
        var rlen = lengths[i],
            ring = new Array(rlen);
        for (var j = 0; j < rlen; j++) {
            var p = new Array(dim);
            for (var d = 0; d < dim && pbf.pos < end; d++) p[d] = transformCoord(pbf.readSVarint());
            ring[j] = p;
        }
        lines[i] = ring;
    }

    return lines;
}

function transformCoord(x) {
    return transformed ? x : x / e;
}


benchmark/benchmark.mjs
import { benchmark } from "@thi.ng/bench";
import fs from "fs";
import path from 'path';
import Pbf from "pbf";
import { fileURLToPath } from 'url';
import { deserialize, serialize } from "v8";

import * as topobuf from "../index.js";


const __filename = fileURLToPath(import.meta.url);

// Create serialized data to use in benchmark
const json = JSON.parse(fs.readFileSync(path.dirname(__filename) + "/fixtures/de-census-blocks.json"));
fs.writeFileSync(path.dirname(__filename) + "/fixtures/de-census-blocks.pbf", topobuf.encode(json, new Pbf()));
fs.writeFileSync(path.dirname(__filename) + "/fixtures/de-census-blocks.buf", serialize(json));

function getJSON(name) {
  return JSON.parse(fs.readFileSync(path.dirname(__filename) + "/fixtures/" + name));
}

// Benchmark tests
function runDecode() {
  topobuf.decode(new Pbf(fs.readFileSync(path.dirname(__filename) + "/fixtures/de-census-blocks.pbf")));
}

function runDecodeV8() {
  deserialize(fs.readFileSync(path.dirname(__filename) + "/fixtures/de-census-blocks.buf"));
}

function runParse() {
  getJSON("de-census-blocks.json")
}

benchmark(runDecode, { title: "decode de census blocks (topobuf)", iter: 100, warmup: 5 });
benchmark(runDecodeV8, { title: "decode de census blocks (v8)", iter: 100, warmup: 5 });
benchmark(runParse, { title: "JSON.parse", iter: 100, warmup: 5 });

encode.js
'use strict';

module.exports = encode;

var keys, keysNum, dim, e, transformed,
    maxPrecision = 1e6;

var geometryTypes = {
    'Point': 0,
    'MultiPoint': 1,
    'LineString': 2,
    'MultiLineString': 3,
    'Polygon': 4,
    'MultiPolygon': 5,
    'GeometryCollection': 6
};

function encode(obj, pbf) {
    keys = {};
    keysNum = 0;
    dim = 0;
    e = 1;
    transformed = false;

    analyze(obj);

    e = Math.min(e, maxPrecision);
    var precision = Math.ceil(Math.log(e) / Math.LN10);

    var keysArr = Object.keys(keys);

    for (var i = 0; i < keysArr.length; i++) pbf.writeStringField(1, keysArr[i]);
    if (dim !== 2) pbf.writeVarintField(2, dim);
    if (precision !== 6) pbf.writeVarintField(3, precision);

    pbf.writeMessage(4, writeTopology, obj);

    keys = null;

    return pbf.finish();
}

function analyze(obj) {
    var i, key;

    if (obj.type === 'Topology') {

        for (key in obj) {
            if (key !== 'type' && key !== 'transform' && key !== 'arcs' && key !== 'objects') saveKey(key);
        }
        analyzeMultiLine(obj.arcs);

        for (key in obj.objects) {
            analyze(obj.objects[key]);
        }

    } else {
        if (obj.type === 'Point') analyzePoint(obj.coordinates);
        else if (obj.type === 'MultiPoint') analyzePoints(obj.coordinates);
        else if (obj.type === 'GeometryCollection') {
            for (i = 0; i < obj.geometries.length; i++) analyze(obj.geometries[i]);
        }

        for (key in obj.properties) saveKey(key);
        for (key in obj) {
            if (key !== 'type' && key !== 'id' && key !== 'coordinates' && key !== 'arcs' &&
                key !== 'geometries' && key !== 'properties') saveKey(key);
        }
    }
}

function analyzeMultiLine(coords) {
    for (var i = 0; i < coords.length; i++) analyzePoints(coords[i]);
}

function analyzePoints(coords) {
    for (var i = 0; i < coords.length; i++) analyzePoint(coords[i]);
}

function analyzePoint(point) {
    dim = Math.max(dim, point.length);

    // find max precision
    for (var i = 0; i < point.length; i++) {
        while (Math.round(point[i] * e) / e !== point[i] && e < maxPrecision) e *= 10;
    }
}

function saveKey(key) {
    if (keys[key] === undefined) keys[key] = keysNum++;
}

function writeGeometry(geom, pbf) {
    pbf.writeVarintField(1, geometryTypes[geom.type]);

    var coords = geom.coordinates,
        arcs = geom.arcs;

    if (geom.type === 'Point') writePoint(coords, pbf);
    else if (geom.type === 'MultiPoint') writeLine(coords, pbf, true);
    else if (geom.type === 'LineString') writeLine(arcs, pbf);
    if (geom.type === 'MultiLineString' || geom.type === 'Polygon') writeMultiLine(arcs, pbf);
    else if (geom.type === 'MultiPolygon') writeMultiPolygon(arcs, pbf);
    else if (geom.type === 'GeometryCollection') {
        for (var i = 0; i < geom.geometries.length; i++) pbf.writeMessage(4, writeGeometry, geom.geometries[i]);
    }

    if (geom.id !== undefined) {
        if (typeof geom.id === 'number' && geom.id % 1 === 0) pbf.writeSVarintField(12, geom.id);
        else pbf.writeStringField(11, geom.id);
    }

    if (geom.properties) writeProps(geom.properties, pbf);
    writeProps(geom, pbf, true);
}

function writeTopology(topology, pbf) {
    if (topology.transform) {
        pbf.writeMessage(1, writeTransform, topology.transform);
        transformed = true;
    }

    var names = Object.keys(topology.objects),
        i, j, d;

    for (i = 0; i < names.length; i++) pbf.writeStringField(2, names[i]);
    for (i = 0; i < names.length; i++) {
        pbf.writeMessage(3, writeGeometry, topology.objects[names[i]]);
    }

    var lengths = [],
        coords = [];

    for (i = 0; i < topology.arcs.length; i++) {
        var arc = topology.arcs[i];
        lengths.push(arc.length);

        for (j = 0; j < arc.length; j++) {
            for (d = 0; d < dim; d++) coords.push(transformCoord(arc[j][d]));
        }
    }

    pbf.writePackedVarint(4, lengths);
    pbf.writePackedSVarint(5, coords);

    writeProps(topology, pbf, true);
}

function writeProps(props, pbf, isCustom) {
    var indexes = [],
        valueIndex = 0;

    for (var key in props) {
        if (isCustom) {
            if (key === 'type') continue;
            else if (props.type === 'Topology')  {
                if (key === 'transform' || key === 'arcs' || key === 'objects') continue;
            } else if (key === 'id' || key === 'coordinates' || key === 'arcs' ||
                       key === 'geometries' || key === 'properties') continue;
        }
        pbf.writeMessage(13, writeValue, props[key]);
        indexes.push(keys[key], valueIndex++);
    }
    pbf.writePackedVarint(isCustom ? 15 : 14, indexes);
}

function writeValue(value, pbf) {
    var type = typeof value;

    if (type === 'string') pbf.writeStringField(1, value);
    else if (type === 'boolean') pbf.writeBooleanField(5, value);
    else if (type === 'object') pbf.writeStringField(6, JSON.stringify(value));
    else if (type === 'number') {
       if (value % 1 !== 0) pbf.writeDoubleField(2, value);
       else if (value >= 0) pbf.writeVarintField(3, value);
       else pbf.writeVarintField(4, -value);
    }
}

function writePoint(point, pbf) {
    var coords = [];
    for (var i = 0; i < dim; i++) coords.push(transformCoord(point[i]));
    pbf.writePackedSVarint(3, coords);
}

function writeLine(line, pbf, isMultiPoint) {
    var coords = [];
    populateLine(coords, line, isMultiPoint);
    pbf.writePackedSVarint(3, coords);
}

function writeMultiLine(lines, pbf) {
    var len = lines.length,
        i;
    if (len !== 1) {
        var lengths = [];
        for (i = 0; i < len; i++) lengths.push(lines[i].length);
        pbf.writePackedVarint(2, lengths);
        // TODO faster with custom writeMessage?
    }
    var coords = [];
    for (i = 0; i < len; i++) populateLine(coords, lines[i]);
    pbf.writePackedSVarint(3, coords);
}

function writeMultiPolygon(polygons, pbf) {
    var len = polygons.length,
        i, j;
    if (len !== 1 || polygons[0].length !== 1 || polygons[0][0].length !== 1) {
        var lengths = [len];
        for (i = 0; i < len; i++) {
            lengths.push(polygons[i].length);
            for (j = 0; j < polygons[i].length; j++) lengths.push(polygons[i][j].length);
        }
        pbf.writePackedVarint(2, lengths);
    }

    var coords = [];
    for (i = 0; i < len; i++) {
        for (j = 0; j < polygons[i].length; j++) populateLine(coords, polygons[i][j]);
    }
    pbf.writePackedSVarint(3, coords);
}

function populateLine(coords, line, isMultiPoint) {
    var i, j;
    for (i = 0; i < line.length; i++) {
        if (!isMultiPoint) coords.push(i ? line[i] - line[i - 1] : line[i]);
        else for (j = 0; j < dim; j++) coords.push(transformCoord(line[i][j] - (i ? line[i - 1][j] : 0)));
    }
}

function transformCoord(x) {
    return transformed ? x : Math.round(x * e);
}

function writeTransform(tr, pbf) {
    pbf.writeDoubleField(1, tr.scale[0]);
    pbf.writeDoubleField(2, tr.scale[1]);
    pbf.writeDoubleField(3, tr.translate[0]);
    pbf.writeDoubleField(4, tr.translate[1]);
}
